### PR TITLE
Move access and opening times into offices and organisations

### PR DESCRIPTION
**Pull request by jkempster34**
> Note: This runs a data migration automatically from the `db/migrate` folder. We discussed this, and it seems to be the way we usually perform this kind of migration. This means a user might lose their edit if the migration runs while they are doing work, but we think this is unlikely!

https://trello.com/c/NFkJQYZw

### Remove unused FormObject model
This FormObject model is not used anywhere (since #5965) and can safely be
removed.

It includes properties that were specific to the feature for which it was
introduced.

### Add access and opening time columns to worldwide organisation and office
We want to remove the `access_and_opening_times` table to make it easier to
move the rendering of Worldwide Organisations and Worldwide Offices out of
Whitehall.

Create migrations to add new columns to `worldwide_organisations` and
`worldwide_offices`. Work will follow to use these instead of the
`access_and_opening_times` table.

### Use access and opening times columns
We have added columns to `worldwide_organisations` and `worldwide_offices` so
that we can remove the `access_and_opening_times` table to make it easier to
move the rendering of Worldwide Organisations and Worldwide Offices out of
Whitehall.

This replaces the use of the AccessAndOpeningTimes model with the new
attributes of the WorldwideOrganisations and WorldwideOffices. There is no
difference in functionality with this change.

### Remove AccessAndOpeningTimes model
This class is no longer used, and the table associated with the model is to be
removed.

### Migrate Access and Opening Times data
We have added new columns to Worldwide Organisations and Worldwide Offices so
that we can remove the access and opening times table.

Copy the data over to the new columns. This specifically uses `update_columns`
instead of `update` so that ActiveRecord callbacks are skipped. Some of the
callbacks are not desired here, for example, we do not want to create an audit
trail entry.

The nested AccessAndOpeningtimes class allows us to use the ActiveRecord query
interface despite the fact that the model has been removed from the app.

### Add pending migration to drop the `access_and_opening_times` table
We have added new columns to Worldwide Organisations and Worldwide Offices so
that we can remove the access and opening times table.

Add a pending migration to drop the `access_and_opening_times` table.

---

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

### CODE REVIEW

**Review by floehopper — COMMENTED** [review]
Unless I'm missing something, the first commit ("Remove unused FormObject model") seems quite orthogonal to the other changes in this PR. Although it's quite small, I think it would be better to move it into a separate PR. Also probably worth linking to the commit where it was last being used, which I think is this one: 2669db52389e0210ddfac0537d30aed6c4c12f71.

I found the 3rd commit ("Use access and opening times columns") quite hard to follow - there's a lot going on. While I think the use of the model form is a significant improvement; I wonder if it could've been used to split the commit up, by introducing the model form in one commit without making any changes to the underlying models/database and then changing the model form to use the new models/database in another commit. I haven't thought this through in detail so it may not be practical and in any case I don't think it's worth re-working now; just something worth bearing in mind.

In the 4th commit ("Migrate Access and Opening Times data") I'm concerned that the `AccessAndOpeningTimes` model will have been removed from the codebase *before* the migration runs as per the 6th commit ("Remove AccessAndOpeningTimes model"). I'm not sure what the regular practice is in Whitehall/GDS, but a more typical way to handle this would be to either (a) define the required bits of the `AccessAndOpeningTimes` model class as a nested class within the migration; or (b) avoid depending on the model methods by using SQL to make the database changes. I imagine this probably worked out for you, because you will have run the migration before removing the model class. However, when this PR is merged and the changes are deployed, I _think_ the migration will fail.

For the 5th commit ("Add pending migration to drop the access_and_opening_times table"), I'm not familiar with the idea of the "pending" migrations. Do these get run automatically at some point or is the idea that we'll create a new PR at some point to move the pending migration into the standard migration directory?

While I've made quite a few comments/suggestions, overall this looks good to me and I don't think you definitely need to make any changes. The one thing that's worth checking is that the migration will run OK without the model being in the codebase, so on that basis I'm going to hold off on approving for now.

File: app/views/admin/worldwide_offices/_worldwide_office.html.erb
Comment (jkempster34) [reply app/views/admin/worldwide_offices/_worldwide_office.html.erb:24]: Sorry, this was a compulsion. I've changed it back!

File: app/models/access_and_opening_times_form.rb
Comment (jkempster34) [reply app/models/access_and_opening_times_form.rb:18]: This is me trying to get around the [Rails/SaveBang](https://www.rubydoc.info/gems/rubocop/0.44.1/RuboCop/Cop/Rails/SaveBang) rule.

I can't really think of a "nice" way of doing this without ignoring the rule, so I've done that instead.

**Comment by jkempster34** [comment]
> Unless I'm missing something, the first commit ("Remove unused FormObject model") seems quite orthogonal to the other changes in this PR. Although it's quite small, I think it would be better to move it into a separate PR. Also probably worth linking to the commit where it was last being used, which I think is this one: [2669db5](https://github.com/alphagov/whitehall/commit/2669db52389e0210ddfac0537d30aed6c4c12f71).
> 
> I found the 3rd commit ("Use access and opening times columns") quite hard to follow - there's a lot going on. While I think the use of the model form is a significant improvement; I wonder if it could've been used to split the commit up, by introducing the model form in one commit without making any changes to the underlying models/database and then changing the model form to use the new models/database in another commit. I haven't thought this through in detail so it may not be practical and in any case I don't think it's worth re-working now; just something worth bearing in mind.
> 
> In the 4th commit ("Migrate Access and Opening Times data") I'm concerned that the `AccessAndOpeningTimes` model will have been removed from the codebase _before_ the migration runs as per the 6th commit ("Remove AccessAndOpeningTimes model"). I'm not sure what the regular practice is in Whitehall/GDS, but a more typical way to handle this would be to either (a) define the required bits of the `AccessAndOpeningTimes` model class as a nested class within the migration; or (b) avoid depending on the model methods by using SQL to make the database changes. I imagine this probably worked out for you, because you will have run the migration before removing the model class. However, when this PR is merged and the changes are deployed, I _think_ the migration will fail.
> 
> For the 5th commit ("Add pending migration to drop the access_and_opening_times table"), I'm not familiar with the idea of the "pending" migrations. Do these get run automatically at some point or is the idea that we'll create a new PR at some point to move the pending migration into the standard migration directory?
> 
> While I've made quite a few comments/suggestions, overall this looks good to me and I don't think you definitely need to make any changes. The one thing that's worth checking is that the migration will run OK without the model being in the codebase, so on that basis I'm going to hold off on approving for now.

I didn't even consider that the migration would fail when deployed, so thanks for raising that. I've used a nested class as you suggested which seems to work fine. 

["Pending" migrations are explained here](https://github.com/alphagov/whitehall/blob/main/db/pending_migration_cleanups/README.txt)

Sorry that the 3rd commit is a bit large, I found it difficult to split up without making it too granular and hard to follow in the history!

**Comment by floehopper** [comment]
> I didn't even consider that the migration would fail when deployed, so thanks for raising that. I've used a nested class as you suggested which seems to work fine.

Cool 👍 

> ["Pending" migrations are explained here](https://github.com/alphagov/whitehall/blob/main/db/pending_migration_cleanups/README.txt)

I saw that, but it was unclear to me how/when those pending migrations are run. Presumably it's a completely manual process of opening a new PR...? Anyway, I'm sure all will become clear!

> Sorry that the 3rd commit is a bit large, I found it difficult to split up without making it too granular and hard to follow in the history!

No worries - it's definitely a challenge!

Given you've addressed the problem with the migration, I'm going to approve this, although it looks as if you've got some conflicts to resolve before merging!

**Review by ollietreend — APPROVED** [review]
This looks good to me. Nice work refactoring and simplifying this!

I'd ordinarily be concerned about releasing view changes & a DB migration all at once – because there could be a brief period of time where the app is expecting the table to have the new structure & data despite it not having been migrated yet.

However, having looked at the number of records we're dealing with (less than 1000), I think it's probably not something worth worrying about on this occasion.

**Pull request merged**
